File: js/problem6.js
/**
 * @param {string} s
 * @param {number} numRows
 * @return {string}
 */
var convert = function(s, numRows) {
    if (numRows === 0) {
        return "";
    }
    
    if ((numRows === 1) || (s.length <= numRows)) {
        return s;
    }
    
    var matrix = [];
    for (let i = 0; i < numRows; i++) {
        matrix.push([]);
    }
    
    let down = false;
    let row = 0;
    
    for (let i = 0; i < s.length; i++) {
        if ((row === 0) || (row === numRows - 1)) {
            down = !down;
        }
        
        matrix[row].push(s[i]);
        if (down) {
            row++;
        } else {
            row--;
        }
        
    }
    
    let res = '';
    
    for (let i = 0; i < numRows; i++) {
        for (let j = 0; j < matrix[i].length; j++) {
            res += matrix[i][j];
        }
    }
    
    return res;
};

File: leetcodeJava/src/main/java/Problem6.java
// 6. ZigZag Conversion

class Solution {
    public String convert(String s, int numRows) {
        if ((numRows == 1) || (s.length() <= numRows)) {
            return s;
        }
        
        List<Character>[] matrix = new ArrayList[numRows];
        
        for (int i = 0; i < numRows; i++) {
            matrix[i] = new ArrayList<Character>();
        }
        
        int loc = 0;
        boolean direction = false;
        for (int i = 0; i < s.length(); i++) {
            matrix[loc].add(s.charAt(i));
            if ((loc == 0) || (loc == numRows - 1)) {
                direction = !direction;
            }
            if (direction == true) {
                loc++;
            } else {
                loc--;
            }
        }
        
        StringBuilder res = new StringBuilder();
        for (int i = 0; i < numRows; i++) {
            for (int j = 0; j < matrix[i].size(); j++) {
                res.append(matrix[i].get(j));
            }
        }
        return res.toString();
    }
}